File: sdk/ts/test/sdk.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";

import { getPublicKey, sign, signInDomain, PUBLIC_KEY_SIZE, SIGNATURE_SIZE } from "@bitspark/archon";
import {
  provePossession,
  verifyPossession,
  possessionMessageBytes,
  POSSESSION_SCHEME_TAG,
  seal,
  open,
} from "../src/index.js";

const seed = new Uint8Array(32).fill(0x09);
const pk = getPublicKey(seed);
const enc = (s: string) => new TextEncoder().encode(s);

test("possession: proves, and refuses every substitution", () => {
  const D = "archon/test/pop";
  const nonce = new Uint8Array(16).fill(0xaa);
  const binding = enc("session:1");
  const sig = provePossession(seed, D, nonce, binding);
  assert.equal(verifyPossession(pk, D, nonce, binding, sig), true);
  assert.equal(verifyPossession(pk, D, new Uint8Array(16).fill(0xab), binding, sig), false, "other nonce");
  assert.equal(verifyPossession(pk, D, nonce, enc("session:2"), sig), false, "other binding");
  assert.equal(verifyPossession(pk, "archon/test/other", nonce, binding, sig), false, "other domain");
  assert.equal(verifyPossession(new Uint8Array(PUBLIC_KEY_SIZE), D, nonce, binding, sig), false, "other key");
  const m = possessionMessageBytes(nonce, binding);
  assert.equal(verifyPossession(pk, D, nonce, binding, sign(m, seed)), false, "raw is not a proof");
  assert.equal(verifyPossession(pk, D, nonce, binding, signInDomain(seed, D, nonce)), false, "bare nonce is not a proof");
});

test("possession: short nonce and empty binding are refused", () => {
  const D = "archon/test/pop";
  assert.throws(() => provePossession(seed, D, new Uint8Array(15).fill(0xaa), enc("b")));
  assert.throws(() => provePossession(seed, D, new Uint8Array(16).fill(0xaa), new Uint8Array(0)));
  assert.throws(() => provePossession(seed, "", new Uint8Array(16).fill(0xaa), enc("b")));
  assert.equal(verifyPossession(pk, D, new Uint8Array(15).fill(0xaa), enc("b"), new Uint8Array(SIGNATURE_SIZE)), false);
  assert.equal(verifyPossession(pk, D, new Uint8Array(16).fill(0xaa), new Uint8Array(0), new Uint8Array(SIGNATURE_SIZE)), false);
});

test("possession: the layout is the pinned one", () => {
  const m = possessionMessageBytes(new Uint8Array(16).fill(0x11), enc("ab"));
  const want = new Uint8Array([POSSESSION_SCHEME_TAG, 0x00, 0x10, ...new Uint8Array(16).fill(0x11), 0x00, 0x02, 0x61, 0x62]);
  assert.deepEqual(m, want);
});

test("envelope: seals, opens, and refuses every tamper", () => {
  const D = "archon/test/env";
  const env = seal(seed, D, enc("payload"));
  const o = open(env, D);
  assert.deepEqual(o.pubkey, pk);
  assert.deepEqual(o.payload, enc("payload"));
  assert.equal(open(seal(seed, D, new Uint8Array(0)), D).payload.length, 0, "empty payload allowed");
  assert.throws(() => open(env, "archon/test/other"), /different domain/);
  const mut = (f: (b: Uint8Array) => void) => {
    const b = new Uint8Array(env);
    f(b);
    return b;
  };
  assert.throws(() => open(mut((b) => (b[b.length - 1]! ^= 1)), D), "payload tampered");
  assert.throws(() => open(mut((b) => (b[0] = 0x78)), D), "bad magic");
  assert.throws(() => open(mut((b) => (b[4] = 0x02)), D), "unknown version");
  assert.throws(() => open(mut((b) => (b[6]! ^= 1)), D), "domain bytes altered");
  assert.throws(() => open(env.subarray(0, env.length - 8), D), "truncated");
  assert.throws(() => open(env.subarray(0, 10), D), "far too short");
  const other = seal(new Uint8Array(32).fill(0x0a), D, enc("payload"));
  const sigAt = 4 + 1 + 1 + D.length + PUBLIC_KEY_SIZE;
  assert.throws(() => open(mut((b) => b.set(other.subarray(sigAt, sigAt + SIGNATURE_SIZE), sigAt)), D), "another key's signature");
});

File: sdk/java/src/test/java/dev/bitspark/archon/sdk/SdkTest.java
package dev.bitspark.archon.sdk;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import dev.bitspark.archon.core.Crypto;
import java.io.ByteArrayOutputStream;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import org.junit.jupiter.api.Test;

/**
 * Unit tests for the Java sdk.
 *
 * <p>The oracle in vectors/sdk.json is the cross-language authority and is driven by
 * conformance/check-java-sdk.mjs; these cover the same ground from inside Java, plus what the
 * oracle's {@code ok} / {@code error} protocol cannot express: which exception is thrown, that
 * verification never throws, and the two refusals whose oracle cases carry no genuine signature (a
 * short nonce and an empty binding on the VERIFY side — see the tests below).
 */
final class SdkTest {

  private static final byte[] SEED = range(0, 32);
  private static final byte[] OTHER_SEED = range(1, 32);
  private static final String DOMAIN = "example/pop/v1";
  private static final byte[] NONCE = range(0, 16);
  private static final byte[] BINDING = bytes("channel");
  private static final String ENV_DOMAIN = "example/env/v1";

  // --- possession: the layout ------------------------------------------------------------

  @Test
  void possessionLayoutIsPinned() {
    byte[] want = concat(new byte[] {0x01, 0x00, 0x10}, NONCE, new byte[] {0x00, 0x07}, BINDING);
    assertArrayEquals(want, Possession.messageBytes(NONCE, BINDING));
  }

  @Test
  void possessionLayoutBounds() {
    Possession.messageBytes(new byte[Possession.MIN_NONCE_SIZE], bytes("x"));
    Possession.messageBytes(new byte[Possession.MAX_FIELD_SIZE], new byte[Possession.MAX_FIELD_SIZE]);
    assertThrows(IllegalArgumentException.class, () -> Possession.messageBytes(new byte[15], bytes("x")));
    assertThrows(
        IllegalArgumentException.class,
        () -> Possession.messageBytes(new byte[Possession.MAX_FIELD_SIZE + 1], bytes("x")));
    assertThrows(IllegalArgumentException.class, () -> Possession.messageBytes(NONCE, new byte[0]));
    assertThrows(
        IllegalArgumentException.class,
        () -> Possession.messageBytes(NONCE, new byte[Possession.MAX_FIELD_SIZE + 1]));
  }

  // --- possession: prove and verify ------------------------------------------------------

  @Test
  void possessionRoundTripAndEveryFieldBinds() {
    byte[] pub = Crypto.publicKeyFromSeed(SEED);
    byte[] proof = Possession.prove(SEED, DOMAIN, NONCE, BINDING);
    assertEquals(64, proof.length);
    assertTrue(Possession.verify(pub, DOMAIN, NONCE, BINDING, proof));
    assertFalse(Possession.verify(pub, DOMAIN, range(1, 16), BINDING, proof));
    assertFalse(Possession.verify(pub, DOMAIN, NONCE, bytes("other channel"), proof));
    assertFalse(Possession.verify(pub, "example/pop/v2", NONCE, BINDING, proof));
    assertFalse(Possession.verify(Crypto.publicKeyFromSeed(OTHER_SEED), DOMAIN, NONCE, BINDING, proof));
  }

  @Test
  void possessionIsDeterministic() {
    assertArrayEquals(
        Possession.prove(SEED, DOMAIN, NONCE, BINDING), Possession.prove(SEED, DOMAIN, NONCE, BINDING));
  }

  @Test
  void aRawSignatureOverTheLayoutIsNotAProof() {
    byte[] pub = Crypto.publicKeyFromSeed(SEED);
    byte[] raw = Crypto.sign(SEED, Possession.messageBytes(NONCE, BINDING));
    assertFalse(Possession.verify(pub, DOMAIN, NONCE, BINDING, raw));
  }

  @Test
  void proveRefusesWhatIsNotAProof() {
    assertThrows(IllegalArgumentException.class, () -> Possession.prove(SEED, DOMAIN, new byte[15], BINDING));
    assertThrows(IllegalArgumentException.class, () -> Possession.prove(SEED, DOMAIN, NONCE, new byte[0]));
    assertThrows(IllegalArgumentException.class, () -> Possession.prove(SEED, "", NONCE, BINDING));
    assertThrows(
        IllegalArgumentException.class,
        () -> Possession.prove(Arrays.copyOf(SEED, 31), DOMAIN, NONCE, BINDING));
  }

  /**
   * The oracle's {@code binding-empty-rejected} verify case carries a signature that is not
   * genuine over its own layout, so it is false whether or not the refusal exists. This one IS
   * genuine: the exact bytes the scheme would sign if it allowed an empty binding, signed in the
   * domain. Only the sdk's own check can refuse it.
   */
  @Test
  void verifyRefusesAGenuineSignatureOverAnUnboundLayout() {
    byte[] pub = Crypto.publicKeyFromSeed(SEED);
    byte[] unbound = concat(new byte[] {0x01, 0x00, 0x10}, NONCE, new byte[] {0x00, 0x00});
    byte[] genuine = Crypto.signInDomain(SEED, DOMAIN, unbound);
    assertFalse(Possession.verify(pub, DOMAIN, NONCE, new byte[0], genuine));
  }

  /** The same gap for {@code nonce-15-rejected}: a genuine signature over the 15-byte layout. */
  @Test
  void verifyRefusesAGenuineSignatureOverAShortNonce() {
    byte[] pub = Crypto.publicKeyFromSeed(SEED);
    byte[] shortNonce = new byte[15];
    byte[] layout = concat(new byte[] {0x01, 0x00, 0x0f}, shortNonce, new byte[] {0x00, 0x07}, BINDING);
    byte[] genuine = Crypto.signInDomain(SEED, DOMAIN, layout);
    assertFalse(Possession.verify(pub, DOMAIN, shortNonce, BINDING, genuine));
  }

  @Test
  void verifyIsTotal() {
    byte[] pub = Crypto.publicKeyFromSeed(SEED);
    byte[] proof = Possession.prove(SEED, DOMAIN, NONCE, BINDING);
    assertFalse(Possession.verify(Arrays.copyOf(pub, 31), DOMAIN, NONCE, BINDING, proof));
    assertFalse(Possession.verify(pub, DOMAIN, NONCE, BINDING, Arrays.copyOf(proof, 63)));
    assertFalse(Possession.verify(pub, "", NONCE, BINDING, proof));
    assertFalse(Possession.verify(pub, DOMAIN, null, BINDING, proof));
    assertFalse(Possession.verify(null, DOMAIN, NONCE, BINDING, proof));
    assertFalse(Possession.verify(pub, null, NONCE, BINDING, proof));
    assertFalse(Possession.verify(pub, DOMAIN, NONCE, BINDING, null));
  }

  // --- envelope --------------------------------------------------------------------------

  @Test
  void envelopeLayoutIsPinned() {
    byte[] sealed = Envelope.seal(SEED, ENV_DOMAIN, bytes("payload"));
    byte[] d = bytes(ENV_DOMAIN);
    assertArrayEquals(bytes("arcn"), Arrays.copyOfRange(sealed, 0, 4));
    assertArrayEquals(bytes("arcn"), Envelope.magic());
    assertEquals(Envelope.VERSION, sealed[4]);
    assertEquals(d.length, sealed[5]);
    assertArrayEquals(d, Arrays.copyOfRange(sealed, 6, 6 + d.length));
    int at = 6 + d.length;
    assertArrayEquals(Crypto.publicKeyFromSeed(SEED), Arrays.copyOfRange(sealed, at, at + 32));
    assertArrayEquals(bytes("payload"), Arrays.copyOfRange(sealed, at + 96, sealed.length));
    assertArrayEquals(concat(new byte[] {0x02}, bytes("payload")), Envelope.messageBytes(bytes("payload")));
  }

  @Test
  void envelopeRoundTrip() {
    Envelope.Opened opened = Envelope.open(Envelope.seal(SEED, ENV_DOMAIN, bytes("payload")), ENV_DOMAIN);
    assertArrayEquals(Crypto.publicKeyFromSeed(SEED), opened.pubkey());
    assertArrayEquals(bytes("payload"), opened.payload());
  }

  @Test
  void openedCannotBeAlteredAfterTheCheck() {
    Envelope.Opened opened = Envelope.open(Envelope.seal(SEED, ENV_DOMAIN, bytes("payload")), ENV_DOMAIN);
    opened.payload()[0] ^= 0x01;
    opened.pubkey()[0] ^= 0x01;
    assertArrayEquals(bytes("payload"), opened.payload());
    assertArrayEquals(Crypto.publicKeyFromSeed(SEED), opened.pubkey());
  }

  @Test
  void emptyPayloadIsAllowed() {
    assertEquals(0, Envelope.open(Envelope.seal(SEED, ENV_DOMAIN, new byte[0]), ENV_DOMAIN).payload().length);
  }

  /**
   * A genuine envelope in another domain: an opener that trusted the envelope's own domain field
   * would accept it.
   */
  @Test
  void theVerifierChoosesTheDomain() {
    byte[] sealed = Envelope.seal(SEED, "example/env/other", bytes("payload"));
    IllegalArgumentException refused =
        assertThrows(IllegalArgumentException.class, () -> Envelope.open(sealed, ENV_DOMAIN));
    assertTrue(refused.getMessage().contains("different domain"), refused.getMessage());
  }

  @Test
  void everyTamperIsRefused() {
    byte[] sealed = Envelope.seal(SEED, ENV_DOMAIN, bytes("payload"));
    int d = ENV_DOMAIN.length();
    for (int index : new int[] {0, 4, 5, 6, 6 + d, sealed.length - 1}) {
      byte[] bent = sealed.clone();
      bent[index] ^= 0x01;
      assertThrows(IllegalArgumentException.class, () -> Envelope.open(bent, ENV_DOMAIN), "byte " + index);
    }
    for (int cut : new int[] {0, 3, 5, 6 + d + 31, 6 + d + 95}) {
      byte[] truncated = Arrays.copyOf(sealed, cut);
      assertThrows(IllegalArgumentException.class, () -> Envelope.open(truncated, ENV_DOMAIN), "cut " + cut);
    }
  }

  /**
   * The scheme tags keep the two layouts apart: sign the payload bytes without the envelope's tag
   * and splice that signature in.
   */
  @Test
  void aSignatureWithoutTheEnvelopeTagIsRefused() {
    byte[] sealed = Envelope.seal(SEED, ENV_DOMAIN, bytes("payload"));
    int at = 6 + ENV_DOMAIN.length() + 32;
    byte[] untagged = Crypto.signInDomain(SEED, ENV_DOMAIN, bytes("payload"));
    System.arraycopy(untagged, 0, sealed, at, 64);
    IllegalArgumentException refused =
        assertThrows(IllegalArgumentException.class, () -> Envelope.open(sealed, ENV_DOMAIN));
    assertTrue(refused.getMessage().contains("does not verify"), refused.getMessage());
  }

  @Test
  void sealRefusesABadDomainOrSeed() {
    assertThrows(IllegalArgumentException.class, () -> Envelope.seal(SEED, "", bytes("x")));
    assertThrows(IllegalArgumentException.class, () -> Envelope.seal(SEED, "d".repeat(256), bytes("x")));
    assertThrows(
        IllegalArgumentException.class, () -> Envelope.seal(Arrays.copyOf(SEED, 31), ENV_DOMAIN, bytes("x")));
  }

  // --- helpers ---------------------------------------------------------------------------

  private static byte[] range(int start, int length) {
    byte[] out = new byte[length];
    for (int i = 0; i < length; i++) {
      out[i] = (byte) (start + i);
    }
    return out;
  }

  private static byte[] bytes(String s) {
    return s.getBytes(StandardCharsets.UTF_8);
  }

  private static byte[] concat(byte[]... parts) {
    ByteArrayOutputStream out = new ByteArrayOutputStream();
    for (byte[] p : parts) {
      out.writeBytes(p);
    }
    return out.toByteArray();
  }
}
